### ISSUE TITLE
eCleaner [originally a question about cookie extensions in general]

### ISSUE THREAD

**Issue by AdKiller** [ghacksuserjs/ghacks-user.js#196]
sorry if this is the wrong place to post a question. I read that you HIGHLY recommend using uBlock Origin, uMatrix and a cookie extension. I thought uMatrix was a cookie extension too. Am I missing something? What is the recommended cookie extension?

Thanks

--

EDIT: hijacked: further down we are talking about "eCleaner (Forget Button)" [GitHub](https://github.com/andy-portmen/ecleaner), [AMO](https://addons.mozilla.org/firefox/addon/ecleaner-forget-button/) - _Thorin_

**Comment by Thorin-Oakenpants** [comment]
Not every one want to use uMatrix, but yes, uM can control cookies on a granular level, so it's superior.

I personally used FF set to block all (because this also control dom storage) and Cookie Controller (default deny-all)  because it allowed me to set "Allow as First Party" (which is not a FF exception rule, only "Allow", and "Allow for Session" are). I should have just used uM.

Right now things are a bit messy. I want to recommend NO cookie extension because none of them work with FPI (first party isolation - this may be fixed but its like P3 last time I looked) - https://bugzilla.mozilla.org/show_bug.cgi?id=1381197 , yup still P3

Cookies are also blocked from extensions in PB mode. The Safe Browsing google cookie is also "protected" from extensions (but @fmarier is our champion here - see https://bugzilla.mozilla.org/show_bug.cgi?id=1362834#c26 ). Add in FPI now. And add in the fact that WebExt can't do anything with local storage (which is tied to cookies), and the whole thing is a complete mess.

I do not understand why FF can't let WebExt have access to indexedDB, and ALL cookies in ALL modes and regardless of FPI on a site permission basis, and to local dom storage - as well as allowing WebExt access to their own storage regardless of prefs (because disabling cookies disables dom.storage.enable which kills a lot of WebExt) ... am getting upset now

We have a wiki page for addons https://github.com/ghacksuserjs/ghacks-user.js/wiki/Appendix-B:-Firefox-Add-ons but it's becoming messy with the whole legacy vs Web Ext transition

**Comment by Thorin-Oakenpants** [comment]
tl;dr: I recommend uMatrix for a granular control. If you also block cookies `(network.cookie.cookieBehavior", 2);` like the user.js does by default, then you also need to allow a site exception

**Comment by Thorin-Oakenpants** [comment]
After ~we~ I finish the 55-alpha release, ~we~~ I plan on cleaning up the add-ons wiki a bit. Also, most of this I have been ~~discussing with knowledgeable folks~~ arguing with myself in #188 , so I'll close this, and any more q's about cookies, join in on that issue :)

**Comment by earthlng** [comment]
> as well as allowing WebExt access to their own storage regardless of prefs 

WEs can do that with the [storage API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage/local) (+permission).

> **Although this API is similar to Window.localStorage it is recommended that you don't use Window.localStorage in extension code**. Firefox will clear data stored by extensions using the localStorage API in various scenarios where users clear their browsing history and data for privacy reasons, while data saved using the storage.local API will be correctly persisted in these scenarios.

It's not mozilla's fault when WE devs don't use the right APIs. (fe. eCleaner)

**Comment by Thorin-Oakenpants** [comment]
Well thats good to know. I quite like eCleaner, but it needs some tweaks (note: almost everything it can clean is already empty for me anyway). 

- needs to one click: make the button do the action and have a dropdown for the options
- optional success/error notification
- better error handling
- and now .. it needs to store its settings properly

How hard it is to fork this? Full disclosure: I am not a coder

**Comment by earthlng** [comment]
> How hard it is to fork this?

Dude, give the guy some time to respond :)

> eCleaner would be handy if we need to turn indexeddb on

how so? do WE's have access to indexedDB's?

**Comment by Thorin-Oakenpants** [comment]
> Dude, give the guy some time to respond

No. I want my toys NOW

> how so? do WE's have access to indexedDB's?

eCleaner is a BULK cleaner, no finesse at all. I guess it doesn't need any special API powers to read origins or attributes or tags or whatever, it just bulk cleans

**Comment by earthlng** [comment]
> do WE's have access to indexedDB's?

eCleaner [uses](https://github.com/andy-portmen/ecleaner/blob/master/data/panel/index.js#L20) the browsingData API's [remove](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/browsingData/remove) "method"

https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/browsingData/remove#Browser_compatibility

> 1. **Firefox does not support removal of: 'fileSystems', 'indexedDB', 'localStorage', or 'serverBoundCertificates'.**
> 2. 'removalOptions' does not support 'protectedWeb' or 'extension' as values of 'originTypes'.

**Comment by Thorin-Oakenpants** [comment]
I have no idea "how" it cleans. Are you saying thet eCleaner fails to clean  'indexedDB', 'localStorage', or 'serverBoundCertificates' even though he says it does?

**Comment by Thorin-Oakenpants** [comment]
Closing this. We have two other issues open re strategies etc for controlling persistent storage, namely #188 and #173 to a lesser degree. Meanwhile eCleaner is still listed in the addons sticky for followup
